lib/postgrix_cluster/api.ex

```
@@ -0,0 +1,79 @@
+defmodule PostgrixCluster.API do
+
+  @moduledoc """
+  Provides functions that interact with the root database
+  in a cluster and provisions new database instances.
+  """
+
+  def createDatabase(pid, db_name) do
+    Postgrex.query!(pid, "set transaction isolation level repeatable read", [])
+    Postgrex.query(pid, "CREATE DATABASE #{db_name} WITH OWNER DEFAULT;", [])
+  end
+
+  def createSchema(pid, schema) do
+    Postgrex.query(pid, "CREATE SCHEMA IF NOT EXISTS #{schema};", [])
+  end
+
+  def dropDatabase(pid, db_name) do
+    Postgrex.query!(pid, "SELECT pg_terminate_backend(pg_stat_activity.pid)
+                          FROM pg_stat_activity WHERE pg_stat_activity.datname = \'#{db_name}\'
+                          AND pid <> pg_backend_pid();", [])
+    Postgrex.query!(pid, "DROP DATABASE IF EXISTS #{db_name};", [])
+  end
+
+  def databaseExists(pid, db_name) do
+    case Postgrex.query(pid, "SELECT 1 AS result FROM pg_database WHERE datname='#{db_name}';", []) do
+      {:ok, result} ->
+        result.rows == [[1]]
+      _ -> false
+    end
+  end
+
+  def addVaultRole(pid, db_name, vault_user, vault_password) do
+    Postgrex.query!(pid, "CREATE ROLE #{vault_user} WITH CREATEROLE
+                          INHERIT LOGIN ENCRYPTED PASSWORD \'#{vault_password}\';", [])
+    Postgrex.query!(pid, "GRANT ALL PRIVILEGES ON DATABASE #{db_name} TO
+                          #{vault_user} WITH GRANT OPTION;", [])
+  end
+
+  def dropVaultRole(pid, db_name, vault_user) do
+    Postgrex.query!(pid, "REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA public from #{vault_user};", [])
+    Postgrex.query!(pid, "REVOKE ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public from #{vault_user};", [])
+    Postgrex.query!(pid, "REVOKE ALL PRIVILEGES ON SCHEMA public from #{vault_user};", [])
+    Postgrex.query!(pid, "REVOKE ALL PRIVILEGES ON DATABASE #{db_name} from #{vault_user};", [])
+    Postgrex.query!(pid, "DROP ROLE IF EXISTS #{vault_user}", [])
```

Do we want to skip later operations if earlier ones fail?

---

Yes, that makes sense.

---

Ok, I had a look and I am going to wrap consecutive operations in `Postgrex.transaction/3`. This should roll back if any of the operations fail. The API function itself should return an error if any of the operations encountered errors.